extensions/amazon-lambda-http/runtime/src/main/java/io/quarkus/amazon/lambda/http/LambdaHttpHandler.java

```diff
@@ -46,19 +53,99 @@ public AwsProxyResponse handleRequest(AwsProxyRequest request, Context context)
         try {
-            return nettyDispatch(connection, request);
+            return nettyDispatch(clientAddress, request);
         } catch (Exception e) {
+            log.error("Request Failure", e);
             return new AwsProxyResponse(500, errorHeaders, "{ \"message\": \"Internal Server Error\" }");
-        } finally {
-            connection.close();
         }
 
     }
 
-    private AwsProxyResponse nettyDispatch(VirtualClientConnection connection, AwsProxyRequest request) throws Exception {
+    private class NettyResponseHandler implements VirtualResponseHandler {
+        AwsProxyResponse responseBuilder = new AwsProxyResponse();
+        ByteArrayOutputStream baos;
+        WritableByteChannel byteChannel;
+        final AwsProxyRequest request;
+        CompletableFuture<AwsProxyResponse> future = new CompletableFuture<>();
+
+        public NettyResponseHandler(AwsProxyRequest request) {
+            this.request = request;
+        }
+
+        public CompletableFuture<AwsProxyResponse> getFuture() {
+            return future;
+        }
+
+        @Override
+        public void handleMessage(Object msg) {
+            try {
+                //log.info("Got message: " + msg.getClass().getName());
+
+                if (msg instanceof HttpResponse) {
+                    HttpResponse res = (HttpResponse) msg;
+                    responseBuilder.setStatusCode(res.status().code());
+
+                    if (request.getRequestSource() == AwsProxyRequest.RequestSource.ALB) {
+                        responseBuilder.setStatusDescription(res.status().reasonPhrase());
+                    }
+                    responseBuilder.setMultiValueHeaders(new Headers());
+                    for (String name : res.headers().names()) {
+                        for (String v : res.headers().getAll(name)) {
+                            responseBuilder.getMultiValueHeaders().add(name, v);
+                        }
+                    }
+                }
+                if (msg instanceof HttpContent) {
+                    HttpContent content = (HttpContent) msg;
+                    int readable = content.content().readableBytes();
+                    if (baos == null && readable > 0) {
+                        baos = createByteStream();
+                    }
+                    for (int i = 0; i < readable; i++) {
+                        baos.write(content.content().readByte());
+                    }
+                }
+                if (msg instanceof FileRegion) {
+                    FileRegion file = (FileRegion) msg;
+                    if (file.count() > 0 && file.transferred() < file.count()) {
+                        if (baos == null)
+                            baos = createByteStream();
+                        if (byteChannel == null)
+                            byteChannel = Channels.newChannel(baos);
+                        file.transferTo(byteChannel, file.transferred());
```

Comment:
This should be file.count().

Reply:
ok, that concerns me because my test passes.

Reply:
Actually no, I am completly wrong here. This should be file.transferred(), which will always be 0.

Reply:
@patriot1burke sorry for the noise, I got confused between the parameters the underlying file channel takes and the FileRegion itself.

Reply:
Yeah, I remember why I had it like that. Its supposed to be position, not how many.